File: dev_tools/scratch/parse_html.php
<?php
$html = file_get_contents('c:/xampp/htdocs/task/index_backup_original.html');

$doc = new DOMDocument();
@$doc->loadHTML('<?xml encoding="utf-8" ?>' . mb_convert_encoding($html, 'HTML-ENTITIES', 'UTF-8'), LIBXML_HTML_NOIMPLIED | LIBXML_HTML_NODEFDTD);
$xpath = new DOMXPath($doc);

$headings = $xpath->query('//h2[contains(@class, "accordion-header")]');
$categories = [];

foreach ($headings as $idx => $heading) {
    $headingId = $heading->getAttribute('id');
    $numMatch = [];
    preg_match('/\d+/', $headingId, $numMatch);
    $catNum = isset($numMatch[0]) ? $numMatch[0] : $idx;

    $titleNode = $xpath->query('.//span[contains(@class, "accordion-title")]', $heading)->item(0);
    $catTitle = $titleNode ? trim($titleNode->textContent) : "Category " . $catNum;

    $collapseId = "collapse" . $catNum;
    $itemsDivId = "items" . $catNum;
    $checkCountId = "check-count" . $catNum;

    $catData = [
        'categoryId' => (int)$catNum,
        'headingId' => $headingId,
        'collapseId' => $collapseId,
        'checkCountId' => $checkCountId,
        'itemsDivId' => $itemsDivId,
        'categoryTitle' => $catTitle,
        'items' => []
    ];

    $itemsDiv = $xpath->query('//div[@id="' . $itemsDivId . '"]')->item(0);

    if ($itemsDiv) {
        $formChecks = $xpath->query('.//div[contains(@class, "form-check")]', $itemsDiv);
        foreach ($formChecks as $fc) {
            $input = $xpath->query('.//input[contains(@class, "form-check-input")]', $fc)->item(0);
            if (!$input) continue;
            $checkId = $input->getAttribute('id');
            $checkName = $input->getAttribute('name');

            $link = $xpath->query('.//a', $fc)->item(0);
            $linkId = $link ? $link->getAttribute('id') : "";
            $targetModalId = $link ? str_replace('#', '', $link->getAttribute('data-bs-target')) : "";

            $labelHtml = "";
            if ($link) {
                foreach ($link->childNodes as $child) {
                    $labelHtml .= $doc->saveHTML($child);
                }
            }

            $modalNode = $xpath->query('//div[@id="' . $targetModalId . '"]')->item(0);
            $modalContentHtml = "";
            if ($modalNode) {
                $modalContent = $xpath->query('.//div[contains(@class, "modal-content")]', $modalNode)->item(0);
                if ($modalContent) {
                    foreach ($modalContent->childNodes as $child) {
                        $modalContentHtml .= $doc->saveHTML($child);
                    }
                }
            }

            $catData['items'][] = [
                'id' => $checkId,
                'name' => $checkName,
                'linkId' => $linkId,
                'targetModalId' => $targetModalId,
                'labelHtml' => trim($labelHtml),
                'modalContentHtml' => trim($modalContentHtml)
            ];
        }
    }

    $categories[] = $catData;
}

file_put_contents(
    'c:/xampp/htdocs/task/data/checklist.json',
    json_encode($categories, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT)
);

echo "Successfully exported " . count($categories) . " categories to data/checklist.json\n";

File: dev_tools/scratch/parse_html.py
import re
import json

def parse_html_to_json(html_path):
    with open(html_path, 'r', encoding='utf-8') as f:
        html = f.read()

    # accordion items
    # pattern for accordion items
    accordion_pattern = re.compile(
        r'<div class="accordion-item">(.*?)</div>\s*</div>\s*</div>\s*</div>', 
        re.DOTALL
    )
    
    # We can also parse accordion by matching headings and contents.
    # Let's extract heading, items, and modals for each collapse.
    
    # Alternative: find all accordion-item elements
    # Let's write a robust parser
    import bs4
    soup = bs4.BeautifulSoup(html, 'html.parser')
    
    accordion = soup.find('div', id='accordion')
    if not accordion:
        print("Accordion not found")
        return
        
    categories = []
    
    items = accordion.find_all('div', class_='accordion-item', recursive=False)
    for idx, item in enumerate(items):
        title_el = item.find('span', class_='accordion-title')
        cat_title = title_el.get_text(strip=True) if title_el else f"Category {idx}"
        
        # heading id
        heading = item.find('h2', class_='accordion-header')
        heading_id = heading.get('id', f'heading{idx}') if heading else f'heading{idx}'
        
        collapse = item.find('div', class_='accordion-collapse')
        collapse_id = collapse.get('id', f'collapse{idx}') if collapse else f'collapse{idx}'
        
        items_div = item.find('div', id=lambda x: x and x.startswith('items'))
        items_div_id = items_div.get('id', f'items{idx}') if items_div else f'items{idx}'
        
        category_data = {
            "categoryId": idx,
            "headingId": heading_id,
            "collapseId": collapse_id,
            "itemsDivId": items_div_id,
            "categoryTitle": cat_title,
            "checkCountId": f"check-count{collapse_id.replace('collapse', '')}",
            "items": []
        }
        
        # Find all form-check within this item
        form_checks = items_div.find_all('div', class_='form-check') if items_div else []
        for fc in form_checks:
            input_el = fc.find('input', class_='form-check-input')
            chk_id = input_el.get('id') if input_el else ""
            
            link_el = fc.find('a')
            target_modal_id = link_el.get('data-bs-target', '').replace('#', '') if link_el else ""
            item_label_html = link_el.decode_contents() if link_el else ""
            link_id = link_el.get('id', '') if link_el else ""
            
            # Find the corresponding modal
            modal_el = items_div.find('div', id=target_modal_id)
            modal_html = ""
            if modal_el:
                # Get modal-dialog inner html or modal-content html
                modal_content = modal_el.find('div', class_='modal-content')
                if modal_content:
                    modal_html = modal_content.decode_contents()
            
            category_data["items"].append({
                "id": chk_id,
                "linkId": link_id,
                "targetModalId": target_modal_id,
                "labelHtml": item_label_html,
                "modalContentHtml": modal_html
            })
            
        categories.append(category_data)
        
    with open('c:/xampp/htdocs/task/data/checklist.json', 'w', encoding='utf-8') as f:
        json.dump(categories, f, ensure_ascii=False, indent=2)
        
    print(f"Successfully extracted {len(categories)} categories to checklist.json")

if __name__ == '__main__':
    parse_html_to_json('c:/xampp/htdocs/task/index.html')
